File: src/main/java/ai/MatrixColumn.java
package ai;

import java.awt.Point;
import java.util.Arrays;
import java.util.LinkedList;

public class MatrixColumn {
	
	final boolean[] column;
	private LinkedList<Point> points;
	
	public MatrixColumn(Point p, boolean[] column) {
		points = new LinkedList<>();
		points.add(p);
		this.column = column;
	}
	
	public void add(MatrixColumn mc) {
		points.addAll(mc.points);
	}
	
	public int size() {
		return points.size();
	}
	
	public LinkedList<Point> getPoints() {
		return points;
	}
	
	public boolean equalsColumn(MatrixColumn matrixColumn) {
		return Arrays.equals(column, matrixColumn.column);
	}
	
	@Override
	public boolean equals(Object obj) {
		if (obj.getClass().equals(this.getClass())) {
			MatrixColumn mc = (MatrixColumn) obj;
			return Arrays.equals(column, mc.column) && points.equals(mc.points);
		}
		return super.equals(obj);
	}

	@Override
	public String toString() {
		String s = "";
		for (boolean b: column) {
			s += b ? "X" : "-";
		}
		s += ":";
		for (Point p: points) {
			s += p.x + "," + p.y + " ";
		}
		return s;
	}
}

File: src/main/kotlin/ai/MatrixColumn.kt
package ai

import java.awt.Point
import java.util.LinkedList

data class MatrixColumn(
	private val points: MutableList<Point>,
	val column: BooleanArray
) {

	constructor(p: Point, column: BooleanArray): this(mutableListOf(p), column)

	fun add(mc: MatrixColumn) {
		points.addAll(mc.points)
	}

	fun size(): Int {
		return points.size
	}

	fun getPoints(): LinkedList<Point> {
		return LinkedList(points)
	}

	fun equalsColumn(matrixColumn: MatrixColumn): Boolean {
		return column.contentEquals(matrixColumn.column)
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (javaClass != other?.javaClass) return false

		other as MatrixColumn

		return points == other.points && column.contentEquals(other.column)
	}

	override fun hashCode(): Int {
		var result = points.hashCode()
		result = 31 * result + column.contentHashCode()
		return result
	}
}